libs/utils/src/array_list.c

```diff
@@ -337,31 +702,69 @@ static int celix_arrayList_compareEntries(const void* voidA, const void* voidB,
     return compare(*a, *b);
 }
 
-void celix_arrayList_sortEntries(celix_array_list_t *list, celix_array_list_compare_entries_fp compare) {
-#if defined(__APPLE__)
-    qsort_r(list->elementData, list->size, sizeof(celix_array_list_entry_t), compare, celix_arrayList_compareEntries);
-#else
-    qsort_r(list->elementData, list->size, sizeof(celix_array_list_entry_t), celix_arrayList_compareEntries, compare);
-#endif
+void celix_arrayList_sort(celix_array_list_t *list) {
+    if (list->compareCallback) {
+        celix_arrayList_sortEntries(list, list->compareCallback);
+    }
 }
 
-#if defined(__APPLE__)
-static int celix_arrayList_compare(void *arg, const void * a, const void *b) {
-#else
-static int celix_arrayList_compare(const void * a, const void *b, void *arg) {
-#endif
-    const celix_array_list_entry_t *aEntry = a;
-    const celix_array_list_entry_t *bEntry = b;
+bool celix_arrayList_equals(const celix_array_list_t* listA, const celix_array_list_t* listB) {
+    if (listA == listB) {
+        return true;
+    }
+    if (!listA || !listB) {
+        return false;
+    }
+    if (listA->size != listB->size) {
+        return false;
+    }
+    if (listA->equalsCallback != listB->equalsCallback) {
+        return false;
+    }
+    for (int i = 0; i < listA->size; ++i) {
+        if (!listA->equalsCallback(listA->elementData[i], listB->elementData[i])) {
+            return false;
+        }
+    }
+    return true;
+}
 
-    celix_arrayList_sort_fp sort = arg;
+celix_array_list_t* celix_arrayList_copy(const celix_array_list_t* list) {
+    celix_array_list_create_options_t opts = CELIX_EMPTY_ARRAY_LIST_CREATE_OPTIONS;
+    opts.elementType = list->elementType;
+    opts.equalsCallback = list->equalsCallback;
+    opts.compareCallback = list->compareCallback;
+    opts.removedCallback = list->removedCallback;
+    opts.removedCallbackData = list->removedCallbackData;
+    opts.simpleRemovedCallback = list->simpleRemovedCallback;
+    celix_autoptr(celix_array_list_t) copy = celix_arrayList_createWithOptions(&opts);
+    if (!copy) {
+        return NULL;
+    }
+
+    for (int i = 0; i < celix_arrayList_size(list); ++i) {
+        celix_array_list_entry_t entry = list->elementData[i];
+        celix_status_t status;
+        if (copy->elementType == CELIX_ARRAY_LIST_ELEMENT_TYPE_STRING) {
+            status = celix_arrayList_addString(copy, entry.stringVal);
+        } else if (copy->elementType == CELIX_ARRAY_LIST_ELEMENT_TYPE_VERSION) {
+            status = celix_arrayList_addVersion(copy, entry.versionVal);
+        } else {
+            status = celix_arrayList_addEntry(copy, entry);
```

Comment:
If `CELIX_ARRAY_LIST_ELEMENT_TYPE_UNDEFINED` is not only intended for backward compatibility but also for support user-provided type, then shallow-copy is not sufficient. If it is only used for compatibility and future extension without modification is not allowed, then the current implementation is OK.

Reply:
Correct, UNDEFINED is IMO temp and should be removed before a 3.0.0 release.